godot-cell/src/lib.rs

```diff
@@ -0,0 +1,304 @@
+    /// references.
+    pub fn make_inaccessible<'a: 'b, 'b>(
+        self: Pin<&'a Self>,
+        current_ref: &'b mut T,
+    ) -> Result<InaccessibleGuard<'b, T>, Box<dyn Error>> {
+        InaccessibleGuard::new(&self.get_ref().state, current_ref)
+    }
+
+    /// Returns `true` if there are any mutable or shared references, regardless of whether the mutable
+    /// references are accessible or not.
+    ///
+    /// In particular this means that it is safe to destroy this cell and the value contained within, as no
+    /// references can exist that can reference this cell.
+    ///
+    /// Keep in mind that in multithreaded code it is still possible for this to return true, and then the
+    /// cell hands out a new borrow before it is destroyed. So we still need to ensure that this cannot
+    /// happen at the same time.
+    pub fn is_currently_bound(self: Pin<&Self>) -> bool {
+        let state = self.state.lock().unwrap();
+
+        state.borrow_state.shared_count() > 0 || state.borrow_state.mut_count() > 0
+    }
+}
+
+// SAFETY:
+// `T` is sync so we can return references to it on different threads.
+// Additionally all internal state is synchronized via a mutex, so we wont have race conditions when trying
+// to use it from multiple threads.
+unsafe impl<T: Sync> Sync for GdCell<T> {}
+
+/// Mutable state of the `GdCell`, bundled together to make it easier to avoid deadlocks when locking the
+/// mutex.
+#[derive(Debug)]
+struct CellState<T> {
+    /// Tracking the borrows this cell has. This ensures relevant invariants are upheld.
+    borrow_state: BorrowState,
+
+    /// Current pointer to the value.
+    ///
+    /// This is an `Option` as we must initialize the value after pinning it.
+    ///
+    /// When a reference is handed to a cell to enable re-entrancy, then this pointer is set to that
+    /// reference.
+    ///
+    /// We always generate new pointer based off of the pointer currently in this field, to ensure any new
+    /// references are derived from the most recent `&mut` reference.
+    ptr: Option<NonNull<T>>,
+}
+
+impl<T> CellState<T> {
+    /// Create a new uninitialized state. Use [`initialize_ptr()`](CellState::initialize_ptr()) to initialize
+    /// it.
+    fn new() -> Self {
+        Self {
+            borrow_state: BorrowState::new(),
+            ptr: None,
+        }
+    }
+
+    /// Initialize the pointer if it is `None`.
+    fn initialize_ptr(&mut self, value: &UnsafeCell<T>) {
+        if self.ptr.is_none() {
+            self.set_ptr(NonNull::new(value.get()).unwrap());
+        } else {
+            panic!("Cannot initialize pointer as it is already initialized.")
+        }
+    }
+
+    /// Returns the current pointer. Panics if uninitialized.
+    fn get_ptr(&self) -> NonNull<T> {
+        self.ptr.unwrap()
+    }
+
+    /// Set the current pointer to the new pointer.
+    fn set_ptr(&mut self, new_ptr: NonNull<T>) {
+        self.ptr = Some(new_ptr);
+    }
```

Comment:
See discussion [here](https://github.com/godot-rust/gdext/pull/501#discussion_r1417934882).

```suggestion
    ptr: *mut T,
}

impl<T> CellState<T> {
    /// Create a new uninitialized state. Use [`initialize_ptr()`](CellState::initialize_ptr()) to initialize
    /// it.
    fn new() -> Self {
        Self {
            borrow_state: BorrowState::new(),
            ptr: std::ptr::null_mut(),
        }
    }

    /// Initialize the pointer if it is `null`.
    fn initialize_ptr(&mut self, value: &UnsafeCell<T>) {
        assert!(self.ptr.is_null(), "pointer is already initialized.");
        self.ptr = value.get();
    }

    /// Returns the current pointer. Panics if uninitialized.
    fn get_ptr(&self) -> NonNull<T> {
        NonNull::new(self.ptr).expect("pointer is uninitialized.")
    }

    /// Set the current pointer to the new pointer.
    fn set_ptr(&mut self, new_ptr: NonNull<T>) {
        self.ptr = new_ptr.as_ptr();
    }
```